test/e2e/tests/subscription-uc.js

```diff
@@ -0,0 +1,106 @@
+
+function generateEmail() {
+    return 'keep.' + shortid.generate() + '@mailtrain.org';
+}
+
+async function subscribe(testUserEmail) {
+    const subscription = {
+        email: testUserEmail
+    };
+
+    await step('User navigates to list subscription page', async () => {
+        await page.webSubscribe.navigate();
+    });
+
+    await step('User submits a valid email', async () => {
+        await page.webSubscribe.enterEmail(testUserEmail);
+        await page.webSubscribe.submit();
+    });
+
+    await step('System shows a notice that further instructions are in the email', async () => {
+        await page.webConfirmSubscriptionNotice.waitUntilVisible();
+    });
+
+    await step('System sends an email with a link to confirm the subscription', async () => {
+        await page.mailConfirmSubscription.fetchMail(testUserEmail);
+    });
+
+    await step('User clicks confirm subscription in the email', async () => {
+        await page.mailConfirmSubscription.click('confirmLink');
+    });
+
+    await step('System shows a notice that subscription has been confirmed', async () => {
+        await page.webSubscribedNotice.waitUntilVisible();
+    });
+
+    await step('System sends an email with subscription confirmation', async () => {
+        await page.mailSubscriptionConfirmed.fetchMail(testUserEmail);
+        subscription.unsubscribeLink = await page.mailSubscriptionConfirmed.link('unsubscribeLink');
+        subscription.manageLink = await page.mailSubscriptionConfirmed.link('manageLink');
+    });
+
+    return subscription;
+}
+
+describe('Subscription use-cases', function() {
+    before(() => driver.manage().deleteAllCookies());
+
+    useCase('Subscription to a public list (main scenario)', async () => {
+        await subscribe(generateEmail());
+    });
+
+    useCase('Subscription to a public list (invalid email)', async () => {
+        await step('User navigates to list subscribe page', async () => {
+            await page.webSubscribe.navigate();
+        });
+
+        await step('User submits an invalid email', async () => {
+            await page.webSubscribe.enterEmail('foo@bar.nope');
+            await page.webSubscribe.submit();
+        });
+
+        await step('System shows a flash notice that email is invalid', async () => {
+            await page.webSubscribe.waitForFlash();
+            expect(await page.webSubscribe.getFlash()).to.contain('Invalid email address');
+        });
+    });
+
+    useCase('Unsubscription from list #1 (one-step, no form)', async () => {
+        const subscription = await subscribe(generateEmail());
+
+        await step('User clicks the unsubscribe button', async () => {
+            await page.mailSubscriptionConfirmed.click('unsubscribeLink');
+        });
+
+        await step('System show a notice that confirms unsubscription', async () => {
+            await page.webUnsubscribedNotice.waitUntilVisible();
```

Comment:
In hindsight, I'm not sure whether using elements for the catch all `waitUntilVisible` was the right decision, at least for the subscription process. Perhaps it would make more sense to [wait for](https://seleniumhq.github.io/selenium/docs/api/javascript/module/selenium-webdriver/lib/until.html) unique text, e.g. "Subscription Confirmed". This way we'd also get an essential text assertion for free and the links/buttons are tested no later than we're trying to click them.

> What do you think? I would possibly do both since they address slightly different cases.

Sounds good. `waitUntilVisibleAfterRefresh ` would solve this problem nicely. Waiting for multiple items makes sense too, yet (for this particular issue) there aren't many readily identifiable elements in the HTML generated by MJML.